Spot nav

jimmytyyang opened the pull request:
## Motivation and Context

<!--- Why is this change required? What problem does it solve? --> Spot navigation in FP

jimmytyyang commented:
Spot robot navigation
(1) Backward action for the oracle skill
(2) Do not need to point towards the object when arriving the grasping point
(3) Collision check for picking/placing locations given robot's navmesh

Example
https://user-images.githubusercontent.com/55121504/235453593-358732d8-74ec-47d4-8041-fb24c16d20f6.mp4

Skylion007 commented on habitat-lab/habitat/tasks/rearrange/actions/oracle_nav_action.py:
```diff
@@ -267,3 +269,310 @@ def step(self, *args, is_last_action, **kwargs):
+                self._config.navmesh_offset_check,
+            )
+            if np.isnan(start_pos).any():
+                print("start_pos contains NaN @oracle_nav_spot_action.py.")
```
Should use logger

eundersander commented on habitat-lab/habitat/tasks/rearrange/utils.py:
```diff
@@ -431,7 +437,53 @@ def place_robot_at_closest_point(
+            # Update the transformation of the agent
+            sim.articulated_agent.base_pos = center_pos
+            # Get the transformation
+            trans = sim.articulated_agent.sim_obj.transformation
```
I think this line and the line above are a bit wasteful/sloppy. Why move the agent `base_pos` at all? It seems like you're using it as a helper to construct the transformation you want. Could you do something like:
```
trans = mn.Matrix4(sim.articulated_agent.sim_obj.transformation)  # make a copy of agent trans
for center_pos in center_pos_list:
    trans.translation = center_pos
    cur_pos = [trans.transform_point(...) ...]
```

jimmytyyang replied:
thanks, this is fixed!

eundersander commented on habitat-lab/habitat/tasks/rearrange/utils.py:
```diff
@@ -431,7 +437,53 @@ def place_robot_at_closest_point(
+                np.array([xz[0], cache_pos[1], xz[2]]) for xz in cur_pos
+            ]
+            is_navigable = [
+                sim.pathfinder.is_navigable(pos) for pos in cur_pos
```
Again, it would be nice to early out on the first failed `pathfinder.is_navigable`.

jimmytyyang replied:
thanks!

eundersander reviewed (commented):
I left some comments. Good progress!

jimmytyyang commented on habitat-baselines/habitat_baselines/rl/hrl/skills/oracle_nav.py:
```diff
@@ -44,10 +44,14 @@ def __init__(
+            self._oracle_nav_ac_idx, _ = find_action_range(
+                action_space, "oracle_nav_action"
+            )
+        except KeyError:
```
Thank you! I added a fix for this.

jimmytyyang commented on habitat-lab/habitat/config/default_structured_configs.py:
```diff
@@ -326,6 +326,47 @@ class OracleNavActionConfig(ActionConfig):
 
 
+@dataclass
+class OracleNavSpotActionConfig(ActionConfig):
```
Thank you! How about "OracleNavWithBackUpActionConfig"?

jimmytyyang commented on habitat-lab/habitat/tasks/rearrange/actions/oracle_nav_action.py:
```diff
@@ -267,3 +269,310 @@ def step(self, *args, is_last_action, **kwargs):
+            np.array([xz[0], self._sim.articulated_agent.base_pos[1], xz[2]])
+            for xz in cur_pos
+        ]
+        is_navigable = [
```
thanks!

jimmytyyang commented on habitat-lab/habitat/tasks/rearrange/actions/oracle_nav_action.py:
```diff
@@ -267,3 +269,310 @@ def step(self, *args, is_last_action, **kwargs):
                 )
+
+
+class velocity_control:
```
thank you!

jimmytyyang commented on habitat-lab/habitat/tasks/rearrange/actions/oracle_nav_action.py:
```diff
@@ -267,3 +269,310 @@ def step(self, *args, is_last_action, **kwargs):
+            vel = OracleNavAction._compute_turn(
+                rel_pos, self._config.turn_velocity, robot_forward
+            )
+            self._sim = vc.act(self._sim, vel)
```
I expose the control parameter, thank you!

jimmytyyang commented on habitat-lab/habitat/tasks/rearrange/multi_task/pddl_sim_state.py:
```diff
@@ -183,8 +183,11 @@ def set_state(
-            agent.base_pos = start_pos
-            agent.base_rot = start_rot
+            # We only teleport the agent when it is allowed
+            if sim_info.enable_teleport_agent_pos:
```
@eundersander  Thank you for the comments. I think this should be fine here. Also, set_state also needs to set the grasping function, so it will not make it a no-op. But @ASzot could also commend on this, thank you!

ASzot replied:
I agree with Eric that this is very risky and I don't like changing this part of the code that all other systems rely on. I also don't think this needs to be a special case. 

I don't understand why Spot does not want to have the teleport condition in the PDDL? If you don't want Spot to teleport for navigation then you should use the "oracle" nav instead of the "teleport" nav or set `apply_postconds: False` for the navigation skill.

jimmytyyang replied:
@ASzot Thanks for this. The reason for not doing teleporting for Spot is because that Spot would be too closed to the target receptacle and thus prone to have collision.

Example of this 
https://user-images.githubusercontent.com/55121504/236859124-a8fdbc45-ab09-40c1-bd1c-6892b7aeda76.mp4

You can see that Spot navigates to the target receptacle and then was teleported. 

Based on 60 testing episodes, I observed that by removing this post-condition, we can reduce the collision rate from 0.36% to 0.11%, which is over 2 times smaller.

I have been already using `oracle_nav` (`nav_to_obj`). But when I turn off `apply_postconds`, it seems that the robot is not navigated anymore. See the video below

https://user-images.githubusercontent.com/55121504/236860287-a40a6702-dd63-4856-b712-bf81fe006033.mp4

I will work this a bit, thanks!

jimmytyyang commented:
@eundersander Hi Eric, thanks for the review. I addressed all the issues, especially for teleporting the robot. Now everything is using trans copy method you suggested. thank you!

Skylion007 commented on habitat-lab/habitat/tasks/rearrange/actions/oracle_nav_action.py:
```diff
@@ -267,3 +267,299 @@ def step(self, *args, is_last_action, **kwargs):
+        robot_pos = np.array(self.cur_articulated_agent.base_pos)
+
+        if curr_path_points is None:
+            raise Exception
```
This seems like a bug. You probably want to raise an instationed instance not a class. And probably should be a RuntimeError

eundersander reviewed (commented):
Another minor comment and let's wait to hear from Andrew about the PDDL bit.

jimmytyyang commented on SIRO_README.md:
```diff
@@ -27,7 +27,8 @@ FP is actually four distinct pieces: Floorplanner scenes, Amazon-Berkeley object
 2. Download [Amazon and Google object archives](https://drive.google.com/drive/folders/1x6i3sDYheCWoi59lv27ZyPG4Ii2GhEZB)
 3. Download FP episodes: [floorplanner.zip](https://drive.google.com/file/d/1Guxn6v2SC5kAtouwMs1DkBJMK3WW0die/view?usp=sharing)
-4. Extract these into `habitat-lab/data` as follows: 
+3-1. (May 1 2023 update): [floorplanner.zip](https://drive.google.com/file/d/1MqhYVnnry3zHj3oqsqbUqy29aAJXz-C5/view?usp=sharing) We generate 100 episodes (nav-pick-place for two objects) using 18 scenes. This dataset is called `s_minitrain.json.gz`
```
Thanks! I removed the old version

ASzot commented on habitat-baselines/habitat_baselines/config/rearrange/rl_hierarchical_oracle_nav_spot_fp.yaml:
```diff
@@ -9,16 +9,32 @@
+        # The navmesh offset for the agent placement
+        navmesh_offset_for_agent_placement: [[0.0, 0.0], [0.25, 0.0], [-0.25, 0.0]]
+        enable_lateral_move: False
+    robot_at_thresh: 3.0
```
Why are you modifying this?

jimmytyyang replied:
Spot navigation code requires considering the non-cylinder navmesh. Also changing `robot_at_thresh` allows Spot to navigate in the tight space to halt early without getting too closed to the target location.

ASzot commented on habitat-baselines/habitat_baselines/rl/hrl/hl/fixed_policy.py:
```diff
@@ -37,8 +37,10 @@ def _parse_solution_actions(self, solution):
-            # Filter out the correct action for the robot.
+            # Filter out the correct action for the robot
             robot_id = "robot_" + self._agent_name.split("_")[1]
+            if robot_id == "robot_agent":
```
What is this for? Why do we have to deal with this special case?

jimmytyyang replied:
It seems that in the config yaml, the agent is called "robot_agent", but the pddl solution yaml is called "robot_0". Maybe we want to have two solution yamls one for single agent, and the other one for multi-agent?

ASzot replied:
Yes, I think the better thing would be to have two solution yamls.

ASzot commented on habitat-lab/habitat/tasks/rearrange/utils.py:
```diff
@@ -431,7 +437,53 @@ def place_robot_at_closest_point(
         )
 
+    # Get the normal agent position
+    agent_pos = sim.safe_snap_point(target_position)
```
Put all the added logic in the `if` statement for `navmesh_offset`. This code is called very often and there shouldn't be any unnecessary operations if `navmesh_offset` is None.

ASzot commented on habitat-lab/habitat/tasks/rearrange/utils.py:
```diff
@@ -431,7 +437,53 @@ def place_robot_at_closest_point(
+
+    accept_agent_center = []
+
+    if navmesh_offset is not None:
```
Put this in a separate method for cleaner code.

jimmytyyang replied:
Thank you, done!

ASzot reviewed (changes requested):
You should remove `enable_teleport_agent_pos`.

jimmytyyang commented:
> You should remove `enable_teleport_agent_pos`.

Thank you, Andrew! This is done, and I turn off `apply_postcond`.

jimmytyyang commented:
Support multi-agent setup
